src/prefect/tasks.py

```diff
@@ -310,7 +310,9 @@ def __init__(
             Callable[["TaskRunContext", Dict[str, Any]], Optional[str]]
         ] = None,
         cache_expiration: Optional[datetime.timedelta] = None,
-        task_run_name: Optional[Union[Callable[[], str], str]] = None,
+        task_run_name: Optional[
+            Union[Callable[[], str], Callable[[Dict[str, Any]], str], str]
+        ] = None,
```

Comment:
Unfortunately, `Callable[[Dict[str, Any]], str]` is not sufficient to capture the fact that this argument must be named `parameters`. For #16298 I'm going to change this to a protocol:

```
class TaskRunNameCallbackWithParameters(Protocol):
    @classmethod
    def is_callback_with_parameters(cls, callable: Callable[..., str]) -> TypeIs[Self]:
        sig = inspect.signature(callable)
        return "parameters" in sig.parameters

    def __call__(self, parameters: dict[str, Any]) -> str:
        ...
```

The typeguard classmethod suffices to help a type checker distinguish between `Callable[[], str]` and `TaskRunNameCallbackWithParameters`, given a `callable()` check.